### PR TITLE
sys/syscalls: make gettimeofday() implementation optional

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

A follow-up to #17732  to not force 64bit on all syscalls users.

With this PR the users of ztimer64 should be limited to:

Revisited list of apps using `ztimer64`:

`for app in examples/* tests/*; do make -C ${app} info-modules | if grep -q ztimer64; then echo ${app};fi;done`
```
examples/ccn-lite-relay
examples/cord_ep
examples/cord_lc
examples/dtls-echo
examples/dtls-sock
examples/dtls-wolfssl
examples/gcoap_dtls
examples/gnrc_networking_mac
examples/lua_basic
examples/lua_REPL
examples/openthread
examples/posix_select
examples/wasm
tests/cpp11_condition_variable
tests/cpp11_mutex
tests/cpp11_thread
tests/driver_ltc4150
tests/driver_pir
tests/gnrc_gomach
tests/lua_loader
tests/nimble_autoconn_ccnl
tests/pkg_arduino_sdi_12
tests/pkg_flatbuffers
tests/pkg_tensorflow-lite
tests/pkg_tinydtls_sock_async
tests/pkg_wolfcrypt-ed25519-verify
tests/pkg_wolfssl
tests/posix_semaphore
tests/pthread
tests/pthread_barrier
tests/pthread_cleanup
tests/pthread_condition_variable
tests/pthread_cooperation
tests/pthread_flood
tests/pthread_rwlock
tests/pthread_tls
tests/saul_drivers
tests/sema
tests/sntp
tests/sys_fido2_ctap
tests/unittests
tests/ztimer64_msg
```

### Testing procedure

Green Murdock, although should probably wait for #17721 or cherry-pick f7f7a9310b17dad691710b02ef34d0d729b2a8ae to test for missing dependencies.

### Issues/PRs references

Follow up to #17732 
Related to #17721 and #17365

### CODE REVIEW

**Comment by fjmolinas** [comment]
Updating `Makefile.ci` is done in f55727c54b52970a265b23ecbf67d02119cb56d8, since the commits making boards go over board here are because of the REMOVEME commit I'll leave them out!

**Comment by fjmolinas** [comment]
Hmm this needs to be dropped 210a091faf47ce1e0899e6ab577bc9b5a496f280

**Comment by fjmolinas** [comment]
BTW with this PR the users of `ztimer64` when rebasing this pr on #17721.

```
for app in examples/* tests/*; do make -C ${app} info-modules | if grep -q ztimer64; then echo ${app};fi;done
examples/ccn-lite-relay
examples/cord_ep
examples/cord_lc
examples/dtls-echo
examples/dtls-sock
examples/gcoap_dtls
examples/gnrc_networking_mac
examples/openthread
examples/wasm
tests/cpp11_condition_variable
tests/cpp11_mutex
tests/cpp11_thread
tests/driver_ltc4150
tests/driver_pir
tests/gnrc_gomach
tests/nimble_autoconn_ccnl
tests/pkg_arduino_sdi_12
tests/pkg_flatbuffers
tests/pkg_tensorflow-lite
tests/pkg_tinydtls_sock_async
tests/posix_semaphore
tests/pthread
tests/pthread_barrier
tests/pthread_cleanup
tests/pthread_condition_variable
tests/pthread_cooperation
tests/pthread_flood
tests/pthread_rwlock
tests/pthread_tls
tests/saul_drivers
tests/sntp
tests/unittests
tests/ztimer64_msg
```

**Comment by fjmolinas** [comment]
Seems fido2 is also using the 64bit  version

**Review by kfessel — COMMENTED** [review]
I like the result of this PR
but i think we should avoid having an empty `gettimeofday()`
(this may lead to somthing just not working a runtime)

I would like to have no implementation if it does not work (prefer linker error instead of runtime failure) a pseudo-module can than activate a/the implementation and ant the needed timers

another way would be something like a fake implementation (counting it's calls) (this should still yield an error so the application can deal with it)

**Comment by kfessel** [comment]
at first I thought fido2 time driven keys in that case it would need a real gettimofday implementation (that is backed by something rtc-like).
in reality:
But than i looked into our impl -> it is used to track HID_TRANSACTION_TIMEOUT  and this defaults to 500 ms -> it can use a  msec timer and even if not it does not need 64 bit of usec (32 Bit (>1h should be enough in any case) (classic example of: there is a 64Bit Timer why should i store less Bits)

**Comment by fjmolinas** [comment]
OK so there are some external packages that required the `gettimeofday` function, should we add a module enabling this? It could be under `newlib_syscalls_default_gettimeofday` and `native_syscalls_gettimeofday`, the other option is patching those packages, but at the end, it will be the same as providing the function. That being said the resolution for `gettimofday` is not specified so maybe we could add `ZTIMER64_MSEC` as the backend and use `ZTIMER64_USEC` only if already included in the build.

**Comment by kfessel** [comment]
> OK so there are some external packages that required the `gettimeofday` function, should we add a module enabling this? It could be under `newlib_syscalls_default_gettimeofday` and `native_syscalls_gettimeofday`, the other option is patching those packages, but at the end, it will be the same as providing the function. That being said the resolution for `gettimofday` is not specified so maybe we could add `ZTIMER64_MSEC` as the backend and use `ZTIMER64_USEC` only if already included in the build.

i think: 
since this is code outside of RIOT we should avoid modifying (patching) it in this PR -> pseudo module gettimeofday and activate it for these packages (i don't see a reason to prefix that that module name with newlib)

**Comment by fjmolinas** [comment]
>  (i don't see a reason to prefix that that module name with newlib)

There is currently no `syscalls` module, just specific `newlib` one...

**Comment by kfessel** [comment]
> > (i don't see a reason to prefix that that module name with newlib)
> 
> There is currently no `syscalls` module, just specific `newlib` one...

if we want to use usemodule to configure the libc/syscalls in we should add such a thing (imagine writing an app that should run on newlib,picolib, native, all of them would have a different enable gettimeofday())

or just have a non usemodule config option (maybe feature (but asking maribu- he told me that they do not set preprocessor flags))

**Comment by fjmolinas** [comment]
> if we want to use usemodule to configure the libc/syscalls in we should add such a thing (imagine writing an app that should run on newlib,picolib, native, all of them would have a different enable gettimeofday())

Well, honestly I'm a bit lost on how this should be done... some `time.h` implementations seem to use `gettimeoday`, some don't, I'm not sure how this works or not with `picolibc`, I could use some help in modeling it if these is the case because I don't quire understand it.

**Comment by kfessel** [comment]
in https://github.com/picolibc/picolibc/issues/38#issuecomment-624770335 keith stated that picolib use `gettimeofday()` to implement `time()`

how about defining `config_%` or `libc_%` or `syscall_%` < as pseudo_modules and use  `module_`X`_gettimeofday` to enable the gettimeofday implementation

or just have a pseudo module `gettimeofday`

**Comment by fjmolinas** [comment]
> in [picolibc/picolibc#38 (comment)](https://github.com/picolibc/picolibc/issues/38#issuecomment-624770335) keith stated that picolib use `gettimeofday()` to implement `time()`
> 
> how about defining `config_%` or `libc_%` or `syscall_%` < as pseudo_modules and use `module_`X`_gettimeofday` to enable the gettimeofday implementation
> 
> or just have a pseudo module `gettimeofday` if we do not want a new name space

What would be a shame is adding `gettimeofday` just because for example `time.h` is included but the function not used...

**Comment by kfessel** [comment]
> What would be a shame is adding `gettimeofday` just because for example `time.h` is included but the function not used...

you do not need an implementation just because the header is included (we should avoid, but can very well have linker errors)

**Comment by fjmolinas** [comment]
> > What would be a shame is adding `gettimeofday` just because for example `time.h` is included but the function not used...
> 
> you do not need an implementation just because the header is included (we should avoid, but can very well have linker errors)

weird, I'm under the impression that is what failed with `EDHOC-C` package, there was no user of `gettimofday`, but the `time.h` header was included, and that showed a linker error.

**Comment by kfessel** [comment]
seems like wolfssl uses gettimeoday() (and edhoc seems to depend on that)

**Comment by fjmolinas** [comment]
> seems like wolfssl uses gettimeoday() (and edhoc seems to depend on that)

Its not using it by default, default backend is `tinycrypt`, in master if I only uncomment the `gettimeofday` newlib definition:

`BOARD=samr21-xpro make -C tests/pkg_edhoc_c/`
```
/home/francisco/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: /home/francisco/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(lib_a-gettimeofdayr.o): in function `_gettimeofday_r':
gettimeofdayr.c:(.text._gettimeofday_r+0xe): undefined reference to `_gettimeofday'
collect2: error: ld returned 1 exit status
make: *** [/home/francisco/workspace/RIOT/tests/pkg_edhoc_c/../../Makefile.include:727: /home/francisco/workspace/RIOT/tests/pkg_edhoc_c/bin/samr21-xpro/tests_pkg_edhoc_c.elf] Error 1

**Comment by kfessel** [comment]
`process.c` uses `time()` which in turn uses `gettimeofday()` in  `generate_conn_id()`

**Comment by fjmolinas** [comment]
> `process.c` uses `time()` which in turn uses `gettimeofday()` (with newlibc and picolibc) in `generate_conn_id()`

Ahh thanks... that eluded me!

**Comment by fjmolinas** [comment]
> time()

Wait this is not true

> > `process.c` uses `time()` which in turn uses `gettimeofday()` (with newlibc and picolibc) in `generate_conn_id()`
> 
> Ahh thanks... that eluded me!

I was locally using a patched version of edhoc, that I had not upstreamed which is why I couldn't find that function... fixed in https://github.com/RIOT-OS/RIOT/pull/17777

**Comment by fjmolinas** [comment]
@kfessel does this look better?

File: sys/Kconfig [hunk 1]
```diff
@@ -48,6 +48,15 @@ config MODULE_PICOLIBC
     bool "Picolibc"
     depends on HAS_PICOLIBC
 
+config MODULE_LIBC_SYSCALLS_GETTIMEOFDAY
+    bool
+    default y
+    select
```
Comment (fjmolinas) [inline sys/Kconfig:54]:
```suggestion
```

**Comment by fjmolinas** [comment]
I think I'll wait for https://github.com/RIOT-OS/RIOT/pull/17719 and then rebase on https://github.com/RIOT-OS/RIOT/pull/17721.

**Comment by fjmolinas** [comment]
Revisited list of apps using `ztimer64`:

```
examples/ccn-lite-relay
examples/cord_ep
examples/cord_lc
examples/dtls-echo
examples/dtls-sock
examples/dtls-wolfssl
examples/gcoap_dtls
examples/gnrc_networking_mac
examples/lua_basic
examples/lua_REPL
examples/openthread
examples/posix_select
examples/wasm
tests/cpp11_condition_variable
tests/cpp11_mutex
tests/cpp11_thread
tests/driver_ltc4150
tests/driver_pir
tests/gnrc_gomach
tests/lua_loader
tests/nimble_autoconn_ccnl
tests/pkg_arduino_sdi_12
tests/pkg_flatbuffers
tests/pkg_tensorflow-lite
tests/pkg_tinydtls_sock_async
tests/pkg_wolfcrypt-ed25519-verify
tests/pkg_wolfssl
tests/posix_semaphore
tests/pthread
tests/pthread_barrier
tests/pthread_cleanup
tests/pthread_condition_variable
tests/pthread_cooperation
tests/pthread_flood
tests/pthread_rwlock
tests/pthread_tls
tests/saul_drivers
tests/sema
tests/sntp
tests/sys_fido2_ctap
tests/unittests
tests/ztimer64_msg
```

**Comment by fjmolinas** [comment]
@kaspar030 @kfessel all green here, would be good to get this in to reduce ztimer64 users!

File: cpu/native/Makefile.dep [hunk 2]
```diff
@@ -40,9 +40,16 @@ ifneq (,$(filter socket_zep,$(USEMODULE)))
...
+ifneq (,$(filter libc_syscalls_gettimeofday,$(USEMODULE)))
+  USEMODULE += xtimer
+  ifneq (,$(filter ztimer_xtimer_compat,$(USEMODULE)))
+    # requires 64bit timestamps
+    USEMODULE += ztimer64_xtimer_compat
+  endif
+endif
+
+ifneq (,$(filter tinydtls lua wolfssl,$(USEPKG)))
+  USEMODULE += libc_syscalls_gettimeofday
```
Comment (kfessel) [inline cpu/native/Makefile.dep:52]: shouldn't the these be checked before `libc_syscalls_gettimeofday` is checked? (so above l.:43)

File: makefiles/pseudomodules.inc.mk [hunk 3]
```diff
@@ -121,6 +121,10 @@ PSEUDOMODULES += log
 PSEUDOMODULES += log_printfnoformat
 PSEUDOMODULES += log_color
 PSEUDOMODULES += lora
+## @defgroup pseudomodule_libc_syscalls_gettimeofday libc_syscalls_gettimeofday
+## @brief Includes implementation of gettimeofday()
+##
+PSEUDOMODULES += libc_syscalls_gettimeofday
```
Comment (kfessel) [inline makefiles/pseudomodules.inc.mk:127]: would you shorten this to `libc_gettimeofday`? - (I think it would be better if it was shorter)

**Review by kfessel — COMMENTED** [review]
I think the pseudo module could have a shorter name. `libc_gettimeofday`

File: cpu/native/Makefile.dep (hunk 2 above)
Comment (fjmolinas) [reply cpu/native/Makefile.dep:52]: Hmm actually the dependency is in the pkg, I think this can be removed.

Comment (fjmolinas) [reply cpu/native/Makefile.dep:52]: I'll move it to the pkg, and check for native.

File: makefiles/pseudomodules.inc.mk (hunk 3 above)
Comment (fjmolinas) [reply makefiles/pseudomodules.inc.mk]: Done

**Comment by fjmolinas** [comment]
@kfessel changes addressed!

File: sys/Makefile.dep [hunk 4]
```diff
@@ -242,13 +242,18 @@ ifneq (,$(filter newlib,$(USEMODULE)))
...
+        # requires 64bit timestamps
+        USEMODULE += ztimer64_xtimer_compat
+      endif
     endif
-  endif
+ endif
 endif
 
+
+
```
Comment (kfessel) [inline sys/Makefile.dep:256]: two extra empty lines

File: sys/Makefile.dep [hunk 5]
```diff
@@ -242,13 +242,18 @@ ifneq (,$(filter newlib,$(USEMODULE)))
...
-      USEMODULE += ztimer64_xtimer_compat
+    ifneq (,$(filter libc_gettimeofday,$(USEMODULE)))
+      USEMODULE += xtimer
+      ifneq (,$(filter ztimer_xtimer_compat,$(USEMODULE)))
+        # requires 64bit timestamps
+        USEMODULE += ztimer64_xtimer_compat
+      endif
     endif
-  endif
+ endif
```
Comment (kfessel) [inline sys/Makefile.dep:252]: that space should be there

**Review by kfessel — COMMENTED** [review]
two small nitpicks in Makefile.dep other than that this is good to go

File: sys/Makefile.dep (hunk 5 above)
Comment (fjmolinas) [reply sys/Makefile.dep]: done

**Review by kfessel — APPROVED** [review]
green murdock

**Comment by benpicco** [comment]
This breaks esp32 :cry: 
https://github.com/RIOT-OS/Release-Specs/issues/247#issuecomment-1111495894
